### PR TITLE
Change: simplified water region evaluation, removed savegame data

### PR DESCRIPTION
**Pull request by Kuhnovic**
## Motivation / Problem

This is closely related to the recently merged region-based ship pathfinder https://github.com/OpenTTD/OpenTTD/pull/10543

There was some discussion with @JGRennison and @SamuXarick on discord, there were worries about desyncs. In 10543 I've tried to prevent desyncs by saving the "is the region initialized "state of all regions in the savegame. The idea was to synchronize the "initialized-ness" in a bug-compatible way, meaning that if it wasn't initialized on the server then it wouldn't be initialized on the joining client either. But this is what can happen:

- server starts a game
- some action, say terraforming, doesn't properly invalidate water region X. Region X is now incorrectly marked as up-to-date.
- a client joins the game. The client needs to scan / label (i.e. update) all water regions that are labeled as up-to-date. This should put the client's water region cache in sync with the one on the server...
- ... but the cache on the server is corrupted, because the "up-to-dateness" of region X is incorrectly set to true. Region X on the server has a different state than region X on the client. This could leads to desyncs.

The current solution of syncing the up-to-date-state of the regions simply doesn't solve this.

Link to discord: https://discord.com/channels/142724111502802944/1008473233844097104/1194656849237127308

## Description

So I'm taking a slightly more simple and brute force approach: *invalidate a water region on any change*. This includes changes to tiles that might not be entirely relevant to the water regions. This by itself shouldn't really have big impact on performance, region invalidation is just a matter of setting a boolean.

## Limitations

- This approach will invalidate more regions than necessary, which in turn will cause more water regions having to update (re-do their connected component labeling). This will have some impact on performance, but I don't think it will be significant.
- I lowered the savegame version and removed the region data from the savegame. This will render some savegames incompatible, but only ones that were created with recent builds or the nightlies. I can change this if that's not acceptable.
- There's some debug code that's nice for testing, but ultimately has to be removed.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/landscape.cpp
Comment (TrueBrain) [inline src/landscape.cpp]: This include shouldn't be here (should be before safeguards), and shouldnt be `<>`?

File: src/pathfinder/water_regions.cpp
Comment (TrueBrain) [inline src/pathfinder/water_regions.cpp]: TODO, but not a draft?

Generally speaking, a PR can be a draft (with TODOs) or be ready-for-review (without TODOs). Makes it easier for us how to look at code :)

File: src/pathfinder/water_regions.cpp
Comment (TrueBrain) [inline src/pathfinder/water_regions.cpp]: Is this actually meant to go in the code? If so, please increase the debugging level; doesn't sound worthy of 1 :)

File: src/water_cmd.cpp
Comment (TrueBrain) [inline src/water_cmd.cpp]: TODOs

**Review by TrueBrain — COMMENTED** [review]
I have no way of judging this code, but if you say it works, I will approve it. It just does need a bit of cleanup :)

**Comment by PeterN** [comment]
This will definitely need to handle the saveload chunk being removed (unless that happens automagically.)

File: src/pathfinder/water_regions.cpp
Comment (Kuhnovic) [reply src/pathfinder/water_regions.cpp]: Map::allocate had debug level 1 so I used that one, although you could argue that the map is quite a bit more important. I've changed this debug level to 2 and the others to 3.

File: src/pathfinder/water_regions.cpp
Comment (Kuhnovic) [reply src/pathfinder/water_regions.cpp]: My bad, should be fixed now

File: src/landscape.cpp
Comment (Kuhnovic) [reply src/landscape.cpp]: Automatically added by VS, should have paid more attention ;)

File: src/saveload/saveload.h
Comment (PeterN) [inline src/saveload/saveload.h:368]: I don't removing this is a good idea, it would mean saves which have the water region chunks wouldn't be loadable in future versions. I'm not sure they'd be loaded in the current version either.

File: src/pathfinder/water_regions.cpp
Comment (TrueBrain) [inline src/pathfinder/water_regions.cpp]: Please keep this in `water_regions_sl.cpp`. No need to hide it.

**Comment by TrueBrain** [comment]
One last thing, sorry about that: I think we should bump the savegame version. Otherwise you can create a savegame which you cannot load, with a very "corrupted savegame" error in the older version. Better to have it say it is not compatible, I think.

File: src/landscape.cpp
Comment (PeterN) [inline src/landscape.cpp:543]: Looks like you now rely on DoClearSquare() being called when any tile is modified to refresh regions (except for objects). Is this always the case? Industries come to mind.

File: src/pathfinder/water_regions.cpp
Comment (PeterN) [inline src/pathfinder/water_regions.cpp]: When does this condition occur? If never and it's a safety check then this should probably be an assert (or asserting that tile is valid)?

File: src/landscape.cpp
Comment (Kuhnovic) [reply src/landscape.cpp:543]: The reason I use DoClearSquare() is that everything that wants to build something on a tile first has to check if that tile can be cleared before it can "claim it for itself". Industry construction does this too. The only exception I've found so far is placing objects, where tile clearing can't be done. I forgot the exact reason but it's in the comments in the place object command, something to do with NewGRFs.

File: src/pathfinder/water_regions.cpp
Comment (Kuhnovic) [reply src/pathfinder/water_regions.cpp]: I think this is a remnant of the time where the initialization code was poorly timed. I remember that abandoning a game with a map larger than the menu game issues, but they no longer seem to occur. I will remove this check, bounds checking is already compiled into the [] operator in debug.

**Review by TrueBrain — APPROVED** [review]
Up to @PeterN to merge this, if he is also okay with these changes.

Tnx for addressing these issues btw, it is appreciated!

**Pull request merged**
